Easier upnp support

Elnaril opened the pull request:
## What was wrong?
Only a limited number of upnp services are listed to setup portforwarding.
The list is hardcoded, it is not straight forward for users to identify the one used by their routers, and it is currently not possible to provide a specific one to Trinity.
See Trinity Issue https://github.com/ethereum/trinity/issues/1959 for more details

## How was it fixed?
The proposed enhancement is divided in 2 parts:

 1/ In ethereum/upnp-port-forward/client.py:
 - A new method `fetch_add_portmapping_services` retrieves and returns all available services which can map ports.
 - The method `setup_port_map` can now receive a service name as a new argument and uses it first to when it tries to map a port.

 2/ In ethereum/trinity:
 - A new component `UpnpServiceNamesComponent` exposes the new command line positional argument `export-upnp-info`, calls `fetch_add_portmapping_services` and print the results, if any, on the terminal.
 - The component `UpnpComponent` exposes now a new command line optional argument `--upnp-service-name` which allows the user to provide a upnp service name to Trinity. This service name is provided to the function `setup_port_map`.

### To-Do

The PR on the Trinity side is ~coming soon~ https://github.com/ethereum/trinity/pull/2116.

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/upnp-port-forward/blob/master/newsfragments/README.md)

[//]: # (See: https://upnp-port-forward.readthedocs.io/en/latest/contributing.html#pull-requests)
- [ ] Add entry to the [release notes](https://github.com/ethereum/upnp-port-forward/blob/master/newsfragments/README.md)

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](http://www.luvbat.com/uploads/rabbit_in_field_of_flowers__8146604322.jpg)

Elnaril committed:
Easier UPnP Support

Fetch all available services which can map ports

Elnaril committed:
Easier UPnP Support

Elnaril committed:
Add NoPortMapServiceFound exception and update __init__.py

Elnaril commented:
The PR on Trinity side has been submitted: https://github.com/ethereum/trinity/pull/2116

pipermerriam commented on upnp_port_forward/client.py:
```diff
@@ -39,9 +47,19 @@ def setup_port_map(
         raise PortMapFailed("No UPnP devices available")
 
+    if wan_service_name:
+        if not wan_service_name.startswith("WAN"):
```
Can we actually know that all of the WAN services will actually have a name starting with `"WAN"`?

Elnaril replied:
I just saw more emphasis on the 'WAN' in: `--upnp-service-name WANSomethingOrOther` than intended.
But anyway it was more a way to start a discussion on possible checks on the user input.
I'll remove it if we don't need any checks. But do we need this feature at all ? (see my comment here https://github.com/ethereum/trinity/pull/2116#issuecomment-737380888 )

pipermerriam replied:
I think we still want this feature (but without the validation). That makes the library usable even if their device isn't yet supported

pipermerriam commented on upnp_port_forward/client.py:
```diff
@@ -96,16 +154,16 @@ def _find_internal_ip_on_device_network(upnp_dev: upnpclient.upnp.Device) -> str
-    "WANPPPConn1",  # Huawei B528s-23a
-)
+def _get_wan_service(
+    upnp_dev: upnpclient.upnp.Device, wan_service_name: Optional[str]
```
I think that we should look at doing this such that it **either** uses a set of explicitly provided names *or* it falls back to the default names by default. 

My intuition is that combining the provided name with the set of default names is going to be problematic because 1) it doesn't give any guarantees that the port mapping will be setup on the provided service name and 2) it doesn't allow us to use the API in a way that excludes the default list which may be desirable.

So we want to allow explicitely specifying one or more names, and if those are not expicitely provided we fall back to the known set of service names.

pipermerriam commented:
I think we still want this feature (but without the validation). That makes the library usable even if their device isn't yet supported

Sent from ProtonMail mobile

-------- Original Message --------
On Dec 2, 2020, 10:40 AM, Elnaril wrote:

> @Elnaril commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [upnp_port_forward/client.py](https://github.com/ethereum/upnp-port-forward/pull/5#discussion_r534358044):
>
>> @@ -39,9 +47,19 @@ def setup_port_map(
>      if not devices:
>          raise PortMapFailed("No UPnP devices available")
>
> +    if wan_service_name:
> +        if not wan_service_name.startswith("WAN"):
>
> I just saw too much emphasis on the 'WAN' in: --upnp-service-name WANSomethingOrOther.
> But anyway it was more a way to start a discussion on possible checks on the user input.
> I'll remove it if we don't need any checks. But do we need this feature at all ? (see my comment here [ethereum/trinity#2116 (comment)](https://github.com/ethereum/trinity/pull/2116#issuecomment-737380888) )
>
> —
> You are receiving this because you commented.
> Reply to this email directly, [view it on GitHub](https://github.com/ethereum/upnp-port-forward/pull/5#discussion_r534358044), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AAGJHASMM47BPV35KTUXUTTSSZ32PANCNFSM4UJP4IDA).

Elnaril committed:
accepts input of several wan service names

remove service names validation

remove service names export feature

Elnaril committed:
rename wan_service_names into required_service_names to avoid confusion with constant WAN_SERVICE_NAMES

Elnaril committed:
Implement export upnp service names feature

Elnaril commented:
@pipermerriam I have moved the export feature in its own module which can be called directly, I removed the validation on user input, allowed several services names to be passed by the users. Also, the port mapping will be tried only on the services passed by the users. The hardcoded ones will be used as a default if no service name is passed by the users.

pipermerriam reviewed (approved):
Very nice work.  Would you be up for a quick update to the documentation:

https://github.com/ethereum/upnp-port-forward/blob/master/docs/upnp_port_forward.rst

A new section that shows how someone can invoke the export from the command line, and then how they can use that output to explicitly provide a service name.  :rocket: :100:

Elnaril commented:
@pipermerriam Thank you very much for the feedback ! :-) And sure, I'll have a look into the doc asap.

Elnaril committed:
Describe export feature in the documentation

Elnaril commented:
@pipermerriam I updated the documentation with the export feature and how to use the service names.
Also, should I commit the automatically generated docs/upnp_port_forward.tools.rst ?

pipermerriam commented:
>  Also, should I commit the automatically generated docs/upnp_port_forward.tools.rst ?

Yes, that is fine.

pipermerriam commented on upnp_port_forward/tools/export.py:
```diff
@@ -0,0 +1,84 @@
+        upnp_service_names_by_device = fetch_add_portmapping_services()
+        service_names_output = "\n"
+        for service in upnp_service_names_by_device:
+            service_names_output = service_names_output.join(
```
This isn't going to produce the output that i think you are wanting.  In the event that there are multiple service names, the first pass will create one well formed block of text, joining the lines below with a newline.  However, on the second pass through the loop, since it continually uses `service_names_output.join(...)` it will interweave the well formed first one inbetween each of the line below.

I would suggest:

1. writing this as a function that you can pass the `upnp_service_names_by_device` list into, which will allow you to actually write a test for the output.
2. Building up the parts into a list and then joining them all at the end.

Elnaril replied:
>This isn't going to produce the output that i think you are wanting. In the event that there are multiple service names, the first pass will create one well formed block of text, joining the lines below with a newline. However, on the second pass through the loop, since it continually uses service_names_output.join(...) it will interweave the well formed first one inbetween each of the line below.

Thank you for pointing this out ! :sweat_smile: I changed accordingly.

pipermerriam commented on upnp_port_forward/client.py:
```diff
@@ -96,16 +108,14 @@ def _find_internal_ip_on_device_network(upnp_dev: upnpclient.upnp.Device) -> str
 
-def _get_wan_service(upnp_dev: upnpclient.upnp.Device) -> upnpclient.upnp.Service:
-    for service_name in WAN_SERVICE_NAMES:
+    service_names_on_trial = (
```
Another name that might be a bit more intuitive than `trial` might be `candidate_service_names`

pipermerriam reviewed (approved):
if you are up for doing a bit of final cleanup we can get this merged and released.

Elnaril committed:
Fix output text formatting

add unit test for export feature

add doc file for upnp_port_forward.tools package

Elnaril commented:
@pipermerriam  Thank you very much for your review ! :smiley:  I added a unit test for the output.

pipermerriam commented:
@Elnaril greatly appreciate you doing this work.  Holidays slow things down but my guess is I'll get this released early next week.

Elnaril commented:
@pipermerriam I really enjoyed doing it ! Thank you very much for your time and guidance ! :-)
